### PR TITLE
DNM: rgw: get usage rest api now return capacity used by each bucket

### PR DESCRIPTION
**Pull request by chenji-kael**
It`s very useful to let users know how much capacity
they have used in each bucket, Venders also need this
function to charge fee by per bucket
Signed-off-by: Ji Chen insomnia@139.com

### CODE REVIEW

**Comment by mattbenjamin** [comment]
@yehudasa prospectively do you see any issue with this?

**Comment by oritwas** [comment]
@chenji-kael , can you open a tracker issue for this?

**Comment by chenji-kael** [comment]
@oritwas 
OK, FEATURE #16191 created

**Comment by oritwas** [comment]
@chenji-kael , can you add it to commit message:
Fixes: url of the tracker issue

**Comment by chenji-kael** [comment]
@oritwas done

File: src/rgw/rgw_auth_s3.cc [hunk 1]
```diff
@@ -29,8 +29,6 @@ static const char *signed_subresources[] = {
   "torrent",
   "uploadId",
   "uploads",
-  "start-date",
```
Comment (oritwas) [inline src/rgw/rgw_auth_s3.cc]: I think you remove those by mistake?

File: src/rgw/rgw_rados.cc [hunk 2]
```diff
@@ -11288,6 +11288,31 @@ int RGWRados::cls_user_sync_bucket_stats(rgw_obj& user_obj, rgw_bucket& bucket)
...
+{
+  map<string, struct rgw_bucket_dir_header> headers;
+  int r = cls_bucket_head(bucket, RGW_NO_SHARD, headers);
+  if (r < 0) {
+    ldout(cct, 20) << "cls_bucket_header() returned " << r << dendl;
+    return r;
+  }
+  
+  bucket.convert(&entry.bucket);
+
```
Comment (oritwas) [inline src/rgw/rgw_rados.cc]: it will be safer if you zeroed the entry size, size_rounded and count fields

File: src/rgw/rgw_rest_s3.cc [hunk 3]
```diff
@@ -408,14 +417,28 @@ void RGWGetUsage_ObjStore_S3::send_response()
        formatter->dump_int("SuccessfulOps", total_usage.successful_ops);
        formatter->close_section(); // total 
        formatter->close_section(); // user
-     } 
-     formatter->dump_int("TotalBytes", header.stats.total_bytes);
```
Comment (oritwas) [inline src/rgw/rgw_rest_s3.cc]: I would keep those in the summary section

**Comment by chenji-kael** [comment]
@oritwas ok

File: src/rgw/rgw_auth_s3.cc (hunk 1 above)
Comment (chenji-kael) [inline src/rgw/rgw_auth_s3.cc]: @oritwas 
no , start-date and end-date is only parameters， should not be treated as sub-resource ， I introduced this mistake at last commit, so I think it need wipe out.
In fact I think usage should really be put here, what`s your opinion?

**Comment by chenji-kael** [comment]
@oritwas I have update this pr, please check it

File: src/rgw/rgw_rados.cc [hunk 4]
```diff
@@ -11288,6 +11288,31 @@ int RGWRados::cls_user_sync_bucket_stats(rgw_obj& user_obj, rgw_bucket& bucket)
...
+  map<string, struct rgw_bucket_dir_header>::iterator hiter = headers.begin();
+  for (; hiter != headers.end(); ++hiter) {
```
Comment (theanalyst) [inline src/rgw/rgw_rados.cc]: can we use C++11 syle auto for the for loops here

File: src/rgw/rgw_user.cc [hunk 5]
```diff
@@ -87,6 +87,44 @@ int rgw_user_sync_all_stats(RGWRados *store, const rgw_user& user_id)
...
+  do {
+    RGWUserBuckets user_buckets;
+    ret = rgw_read_user_buckets(store, user_id, user_buckets, marker,
+				string(), max_entries, false, &is_truncated);
+    if (ret < 0) {
+      ldout(cct, 0) << "failed to read user buckets: ret=" << ret << dendl;
+      return ret;
+    }
+    map<string, RGWBucketEnt>& buckets = user_buckets.get_buckets();
+    for (map<string, RGWBucketEnt>::iterator i = buckets.begin();
```
Comment (theanalyst) [inline src/rgw/rgw_user.cc]: see earlier comment

File: src/rgw/rgw_rest_s3.cc (hunk 3 above)
Comment (chenji-kael) [inline src/rgw/rgw_rest_s3.cc]: @oritwas 
IMHO, summary is a sum of all data search from usage log. and it represent process datas
but "TotalBytes" "TotalBytesRounded" "TotalEntries" are current capacity used by user, It`s a finally result
So I think put them in "CapacityUsed" is more suitable

Comment (theanalyst) [inline src/rgw/rgw_rest_s3.cc]: I'm not sure whether changing the json response here is a good idea, there might be other consumers who are built on top the rest api. For eg. the openstack ceilometer rgw plugin iirc used the "Summary" bits for quite sometime.. not sure what the state is right now

Comment (chenji-kael) [inline src/rgw/rgw_rest_s3.cc]: @theanalyst 
If really someone else used this api ，I agree to keep it as before。
I will change it back

Comment (theanalyst) [inline src/rgw/rgw_rest_s3.cc]: I'm not sure completley, but iirc the first ceilometer rgw plugin did use the usage api, so might be worth worth asking if there are other consumers of this api as well

**Comment by yehudasa** [comment]
@theanalyst @oritwas is this one ready?

**Comment by oritwas** [comment]
it looks like the comments were not fixed.

**Comment by oritwas** [comment]
also it will be good to join https://github.com/ceph/ceph/pull/9495 with this pull request

**Comment by chenji-kael** [comment]
@oritwas @yehudasa 
in fact, I do not know how to treat this PR now due to  #9495. and most comments here have been fixed except using some C++ 11 feature

**Comment by yehudasa** [comment]
@oritwas @theanalyst @chenji-kael what is the status of this PR?

**Comment by oritwas** [comment]
@chenji-kael, please rebase

File: src/rgw/rgw_rados.cc (hunk 2 above)
Comment (oritwas) [inline src/rgw/rgw_rados.cc]: you need to zero entry values (see my previous comment)

**Comment by chenji-kael** [comment]
@oritwas rebase done

**Comment by yehudasa** [comment]
@oritwas ping

**Comment by cbodley** [comment]
compression support was added to radosgw recently, and `struct rgw_bucket_category_stats` now includes an `actual_size` field that represents the compressed size in rados. we might consider exposing that information also

File: src/rgw/rgw_user.cc [hunk 6]
```diff
@@ -83,6 +83,44 @@ int rgw_user_sync_all_stats(RGWRados *store, const rgw_user& user_id)
...
+      marker = i->first;
+
+      RGWBucketEnt& bucket_ent = i->second;
+      cls_user_bucket_entry entry;
+      ret = store->cls_user_get_bucket_stats(bucket_ent.bucket, entry);
+      if (ret < 0) {
+        ldout(cct, 0) << "ERROR: could not get bucket stats: ret=" << ret << dendl;
+        return ret;
+      }
+      buckets_usage_map.insert(pair<string, cls_user_bucket_entry>(bucket_ent.bucket.name, entry));
```
Comment (cbodley) [inline src/rgw/rgw_user.cc]: `buckets_usage_map.emplace(bucket_ent.bucket.name, entry);` should work here

**Comment by oritwas** [comment]
jenkins test this please

**Comment by dachary** [comment]
jenkins test this please (asok)

**Comment by liewegas** [comment]
@oritwas what should we do with this pr?

**Comment by oritwas** [comment]
The changes needed are small.
I will create a new pr based on this one and add the needed changes.

**Comment by oritwas** [comment]
opened #14287 with the fixes

**Pull request closed without merging**